pkg/fqdn/service/cell.go

```diff
@@ -0,0 +1,105 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package service
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+
+	"github.com/cilium/hive/cell"
+	"github.com/cilium/hive/job"
+	"github.com/spf13/pflag"
+
+	"github.com/cilium/cilium/pkg/endpointmanager"
+	"github.com/cilium/cilium/pkg/fqdn/messagehandler"
+	"github.com/cilium/cilium/pkg/ipcache"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/time"
+)
+
+// Cell provides the standalone DNS proxy gRPC server.
+// It is responsible for sending the DNS rules and IP cache updates
+// to the standalone DNS proxy. It also handles the DNS responses
+// from the standalone DNS proxy and updates the DNS rules and IP cache
+// accordingly. It receives the DNS rules during the endpoint regeneration
+// event and listens for ip cache updates from the ipcache.
+var Cell = cell.Module(
+	"sdp-grpc-server",
+	"Provides the standalone DNS proxy gRPC server",
+
+	cell.Config(defaultConfig),
+	cell.Provide(newServer),
+)
+
+type serverParams struct {
+	cell.In
+
+	Logger            *slog.Logger
+	EndpointManager   endpointmanager.EndpointManager
+	DNSRequestHandler messagehandler.DNSMessageHandler
+	IPCache           *ipcache.IPCache
+	JobGroup          job.Group
+	Config            FQDNConfig
+	DaemonConfig      *option.DaemonConfig
+}
+
+func newServer(params serverParams) *FQDNDataServer {
+	srv := NewServer(params.EndpointManager, params.DNSRequestHandler, params.Config.StandaloneDNSProxyServerPort, params.Logger)
+
+	if !params.Config.EnableStandaloneDNSProxy {
+		return srv
+	}
+
+	if !params.DaemonConfig.EnableL7Proxy {
+		srv.log.Error("Standalone DNS proxy requires L7 proxy to be enabled")
+		return srv
+	}
+
+	if params.DaemonConfig.ToFQDNsProxyPort == 0 || params.Config.StandaloneDNSProxyServerPort == 0 {
+		srv.log.Error("Standalone DNS proxy requires a valid port number to be set")
+		return srv
+	}
+
+	params.IPCache.AddListener(srv)
+
+	params.JobGroup.Add(job.OneShot("sdp-grpc-server", func(ctx context.Context, health cell.Health) error {
+		health.OK(fmt.Sprintf("Serving at %d", params.Config.StandaloneDNSProxyServerPort))
```

Comment:
Wouldn't it make more sense to set the health as OK once we are actually serving?

Reply:
I don't think it matters that much as it's to do that we'd need to actually make a test call to the gRPC server. If it is failing to serve the job would stop and the health status would anyway get updated with an error. But yeah, we do need to handle error from `ListenAndServe`.